Out of bounds panic on YAML parsing

This bug was privately disclosed and told to open here.

**What happened**:

NewYAMLOrJSONDecoder from k8s.io/apimachinery/pkg/util/yaml has an index out of bounds issue which can cause callers to panic.

Download attached reproducer file
Run golang test:
```go
[repro.txt](https://github.com/kubernetes/kubernetes/files/5759584/repro.txt)

func TestCrash(t *testing.T) {
    f, _ := ioutil.ReadFile("repro")
    decoder := kubeyaml.NewYAMLOrJSONDecoder(bytes.NewBuffer(f), 1024)
    for {
        into := &struct{}{}
        err := decoder.Decode(&into)
        if err == io.EOF {
            break
        }
    }
}
```
When run with k8s.io/apimachinery v0.19.3:

```
=== RUN   TestCrash
--- FAIL: TestCrash (0.00s)
panic: runtime error: slice bounds out of range [:524297] with length 1024 [recovered]
    panic: runtime error: slice bounds out of range [:524297] with length 1024

goroutine 26 [running]:
testing.tRunner.func1.1(0x1ac7380, 0xc00090dc00)
    /usr/local/go/src/testing/testing.go:940 +0x2f5
testing.tRunner.func1(0xc000b4a120)
    /usr/local/go/src/testing/testing.go:943 +0x3f9
panic(0x1ac7380, 0xc00090dc00)
    /usr/local/go/src/runtime/panic.go:969 +0x166
k8s.io/apimachinery/pkg/util/yaml.(*YAMLOrJSONDecoder).Decode(0xc000529f30, 0x177b980, 0xc0000106c8, 0x0, 0x0)
    /home/howardjohn/go/pkg/mod/k8s.io/apimachinery@v0.19.3/pkg/util/yaml/decoder.go:245 +0x620
istio.io/istio/tests/fuzz.TestCrash(0xc000b4a120)
    /home/howardjohn/go/src/istio.io/istio/tests/fuzz/regression_test.go:22 +0x12d
testing.tRunner(0xc000b4a120, 0x1cebf78)
    /usr/local/go/src/testing/testing.go:991 +0xdc
created by testing.(*T).Run
    /usr/local/go/src/testing/testing.go:1042 +0x357
```

Additionally, this can be seen in kubectl directly (below is kubectl 1.20.0):

```
$ kubectl apply -f  repro --dry-run=client --validate=false
panic: runtime error: slice bounds out of range [:526607] with length 4096

goroutine 1 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/yaml.(*YAMLOrJSONDecoder).Decode(0xc0005af6d0, 0x1a3d3e0, 0xc00094d7a0, 0x0, 0x0)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/yaml/decoder.go:276 +0x67d
k8s.io/kubernetes/vendor/k8s.io/cli-runtime/pkg/resource.(*StreamVisitor).Visit(0xc000428900, 0xc000997ea0, 0x1000, 0xc0009a5000)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/cli-runtime/pkg/resource/visitor.go:576 +0xb9
k8s.io/kubernetes/vendor/k8s.io/cli-runtime/pkg/resource.(*FileVisitor).Visit(0xc000997da0, 0xc000997ea0, 0x0, 0x0)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/cli-runtime/pkg/resource/visitor.go:528 +0x25d
k8s.io/kubernetes/vendor/k8s.io/cli-runtime/pkg/resource.EagerVisitorList.Visit(0xc0030f8740, 0x1, 0x1, 0xc000428a80, 0xc000428a01, 0xc000428a80)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/cli-runtime/pkg/resource/visitor.go:213 +0xc5
k8s.io/kubernetes/vendor/k8s.io/cli-runtime/pkg/resource.FlattenListVisitor.Visit(0x1cee740, 0xc000997dc0, 0x1cf9780, 0x2a9d9d0, 0xc0009a2180, 0xc000428a00, 0x10000c0030f8730, 0xc0005af8f8)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/cli-runtime/pkg/resource/visitor.go:391 +0x86
k8s.io/kubernetes/vendor/k8s.io/cli-runtime/pkg/resource.FlattenListVisitor.Visit(0x1cee780, 0xc0009a21b0, 0x1cf9780, 0x2a9d9d0, 0xc0009a2180, 0xc000997e60, 0x40d930, 0xc0005af9a8)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/cli-runtime/pkg/resource/visitor.go:391 +0x86
k8s.io/kubernetes/vendor/k8s.io/cli-runtime/pkg/resource.ContinueOnErrorVisitor.Visit(0x1cee780, 0xc0009a2210, 0xc0004289c0, 0xc0030f8701, 0xc0004289c0)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/cli-runtime/pkg/resource/visitor.go:354 +0x9c
k8s.io/kubernetes/vendor/k8s.io/cli-runtime/pkg/resource.DecoratedVisitor.Visit(0x1cee700, 0xc0030f8770, 0xc000997e00, 0x3, 0x4, 0xc0030f8780, 0xc00095d680, 0xc0005afaf8)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/cli-runtime/pkg/resource/visitor.go:326 +0x86
k8s.io/kubernetes/vendor/k8s.io/cli-runtime/pkg/resource.(*Result).Infos(0xc00095d680, 0xc00095d680, 0x0, 0xc0007e0000, 0x2a9d9d0, 0xc0005afbb0)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/cli-runtime/pkg/resource/result.go:122 +0xc2
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/apply.(*ApplyOptions).GetObjects(0xc000696820, 0xc0030f86c0, 0xc0005afbd0, 0x160f02d, 0x1922840, 0xc0030f86c0)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/apply/apply.go:342 +0x185
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/apply.(*ApplyOptions).Run(0xc000696820, 0x0, 0x1bb70f8)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/apply/apply.go:373 +0x5e
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/apply.NewCmdApply.func1(0xc000960000, 0xc00041ac00, 0x0, 0x4)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/apply/apply.go:180 +0x107
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute(0xc000960000, 0xc00041abc0, 0x4, 0x4, 0xc000960000, 0xc00041abc0)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:854 +0x2c2
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc0001c8b00, 0xc000184180, 0xc000184120, 0x6)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:958 +0x375
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute(...)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:895
main.main()
        _output/dockerized/go/src/k8s.io/kubernetes/cmd/kubectl/kubectl.go:49 +0x21d
```

**Environment**:
- Kubernetes version (use `kubectl version`): Verified with 1.19, 1.20

[repro.txt](https://github.com/kubernetes/kubernetes/files/5759585/repro.txt)

--- a/staging/src/k8s.io/apimachinery/pkg/util/yaml/BUILD
+++ b/staging/src/k8s.io/apimachinery/pkg/util/yaml/BUILD
@@ -19,7 +19,6 @@ go_library(
     importpath = "k8s.io/apimachinery/pkg/util/yaml",
     deps = [
         "//staging/src/k8s.io/apimachinery/pkg/util/json:go_default_library",
-        "//vendor/k8s.io/klog/v2:go_default_library",
         "//vendor/sigs.k8s.io/yaml:go_default_library",
     ],
 )

--- a/staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go
@@ -22,13 +22,11 @@ import (
 	"encoding/json"
 	"fmt"
 	"io"
-	"io/ioutil"
 	"strings"
 	"unicode"
 
 	jsonutil "k8s.io/apimachinery/pkg/util/json"
 
-	"k8s.io/klog/v2"
 	"sigs.k8s.io/yaml"
 )
 
@@ -215,16 +213,15 @@ type YAMLOrJSONDecoder struct {
 	bufferSize int
 
 	decoder decoder
-	rawData []byte
 }
 
 type JSONSyntaxError struct {
-	Line int
-	Err  error
+	Offset int64
+	Err    error
 }
 
 func (e JSONSyntaxError) Error() string {
-	return fmt.Sprintf("json: line %d: %s", e.Line, e.Err.Error())
+	return fmt.Sprintf("json: offset %d: %s", e.Offset, e.Err.Error())
 }
 
 type YAMLSyntaxError struct {
@@ -250,35 +247,18 @@ func NewYAMLOrJSONDecoder(r io.Reader, bufferSize int) *YAMLOrJSONDecoder {
 // provide object, or returns an error.
 func (d *YAMLOrJSONDecoder) Decode(into interface{}) error {
 	if d.decoder == nil {
-		buffer, origData, isJSON := GuessJSONStream(d.r, d.bufferSize)
+		buffer, _, isJSON := GuessJSONStream(d.r, d.bufferSize)
 		if isJSON {
 			d.decoder = json.NewDecoder(buffer)
-			d.rawData = origData
 		} else {
 			d.decoder = NewYAMLToJSONDecoder(buffer)
 		}
 	}
 	err := d.decoder.Decode(into)
-	if jsonDecoder, ok := d.decoder.(*json.Decoder); ok {
-		if syntax, ok := err.(*json.SyntaxError); ok {
-			data, readErr := ioutil.ReadAll(jsonDecoder.Buffered())
-			if readErr != nil {
-				klog.V(4).Infof("reading stream failed: %v", readErr)
-			}
-			js := string(data)
-
-			// if contents from io.Reader are not complete,
-			// use the original raw data to prevent panic
-			if int64(len(js)) <= syntax.Offset {
-				js = string(d.rawData)
-			}
-
-			start := strings.LastIndex(js[:syntax.Offset], "\n") + 1
-			line := strings.Count(js[:start], "\n")
-			return JSONSyntaxError{
-				Line: line,
-				Err:  fmt.Errorf(syntax.Error()),
-			}
+	if syntax, ok := err.(*json.SyntaxError); ok {
+		return JSONSyntaxError{
+			Offset: syntax.Offset,
+			Err:    syntax,
 		}
 	}
 	return err

--- a/staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder_test.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder_test.go
@@ -242,8 +242,9 @@ func TestDecodeBrokenJSON(t *testing.T) {
 	if err == nil {
 		t.Fatal("expected error with json: prefix, got no error")
 	}
-	if !strings.HasPrefix(err.Error(), "json: line 3:") {
-		t.Fatalf("expected %q to have 'json: line 3:' prefix", err.Error())
+	const msg = `json: offset 28: invalid character '"' after object key:value pair`
+	if msg != err.Error() {
+		t.Fatalf("expected %q, got %q", msg, err.Error())
 	}
 }
 

